### PR TITLE
man: Update systemd.killMode docs

### PR DESCRIPTION
**Pull request by ClydeByrdIII**
My initial reading of killMode= gave me the impression that killMode=control-group would kill all non-main PID processes when MAINPID exited, but that's not the case in unified cgroup hierarchy. Only killMode=mixed follows the documentation.
In unified cgroup hierarchy, when using default kill settings, all processes are sent sigterm, then after kill timeout any remaining processes are sent sigkill. The current documentation implies that as soon as MAINPID exits all remaining processes are sent sigkill, but that only applies for killMode=mixed.

Aside: There is a fixme in https://github.com/systemd/systemd/blob/main/src/core/unit.c#L4612-L4618
about the legacy cgroup hierarchy following the current documented behavior, but that's more out of buggy cgroupv1 behavior.

### CODE REVIEW

**Comment by yuwata** [comment]
Closing, as the commit is included in #22694.

**Comment by anitazha** [comment]
Clyde split the commit out so I'm reopening this one

**Comment by anitazha** [comment]
Marking as needs feedback to address the questions left at on https://github.com/systemd/systemd/pull/22694#issuecomment-1064324763

**Comment by ClydeByrdIII** [comment]
@poettering  sorry to move the discussion from https://github.com/systemd/systemd/pull/22694#issuecomment-1064324763 to here, but here's a repro

I've crafted a systemd run command that will spawn a bash process that starts a perl process that will ignore sigterm and sleep for 1 hour in the background. The bash process will subsequently sleep for a few seconds to give perl time to set up its signal handling and exit while the perl child process is still sleeping.

When killMode=Mixed the expectation is that when MAINPID (in this case bash) exits, all remaining processes in the cgroup is immediately sent FinalKillSignal, which by default is SIGKILL.
 
```
systemd-run -p "KillMode=mixed" -- /bin/sh -c "setsid perl -e ""'\$SIG{TERM}='IGNORE'; sleep 3600' & sleep 5"
```

results in these systemd logs
```
Thu 2022-03-10 16:42:20 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: About to execute /bin/sh -c 'setsid perl -e '$SIG{TERM}='IGNORE'; sleep 3600' & sleep 5'
...
Thu 2022-03-10 16:42:20 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Changed dead -> running
Thu 2022-03-10 16:42:20 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Executing: /bin/sh -c 'setsid perl -e '$SIG{TERM}='IGNORE'; sleep 3600' & sleep 5'
Thu 2022-03-10 16:42:20 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Failed to send unit change signal for run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Connection reset by peer
Thu 2022-03-10 16:42:20 PST systemd-journald.service: Got notification message from PID 21 (FDSTORE=1)
Thu 2022-03-10 16:42:20 PST systemd-journald.service: Added fd 9 (n/a) to fd store.
Thu 2022-03-10 16:42:20 PST Bus private-bus-connection: changing state CLOSING → CLOSED
Thu 2022-03-10 16:42:20 PST Got disconnect on private connection.
Thu 2022-03-10 16:42:25 PST Received SIGCHLD from PID 16384 (sh).
Thu 2022-03-10 16:42:25 PST Child 16384 (sh) died (code=exited, status=0/SUCCESS)
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Child 16384 belongs to run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Main process exited, code=exited, status=0/SUCCESS (success)
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Killing process 16385 (perl) with signal SIGKILL.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Killing process 16385 (perl) with signal SIGKILL.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Deactivated successfully.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Unit process 16385 (perl) remains running after unit stopped.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Service will not restart (restart setting)
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Changed running -> dead
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Failed to destroy cgroup /task/system.slice/run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service, ignoring: Device or resource busy
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Consumed 7ms CPU time.
Thu 2022-03-10 16:42:25 PST run-r0adde4bbbdaf4d2e9cbb78ad485bc523.service: Collecting.
Thu 2022-03-10 16:42:25 PST Child 16385 (perl) died (code=killed, status=9/KILL)
```
You can see as soon as bash dies, everything is sigkilled immediately.

So mixed behaves according to the documentation.

But for killMode=control-group it's not what I interpreted the documentation:

```
systemd-run -p "KillMode=control-group" -- /bin/sh -c "setsid perl -e ""'\$SIG{TERM}='IGNORE'; sleep 3600' & sleep 5"
```
outputs
```
Thu 2022-03-10 16:51:50 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Executing: /bin/sh -c 'setsid perl -e '$SIG{TERM}='IGNORE'; sleep 3600' & sleep 5'
Thu 2022-03-10 16:51:50 PST systemd-journald.service: Got notification message from PID 21 (FDSTORE=1)
Thu 2022-03-10 16:51:50 PST systemd-journald.service: Added fd 9 (n/a) to fd store.
Thu 2022-03-10 16:51:50 PST Bus private-bus-connection: changing state CLOSING → CLOSED
Thu 2022-03-10 16:51:50 PST Got disconnect on private connection.
Thu 2022-03-10 16:51:55 PST Received SIGCHLD from PID 16406 (sh).
Thu 2022-03-10 16:51:55 PST Child 16406 (sh) died (code=exited, status=0/SUCCESS)
Thu 2022-03-10 16:51:55 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Child 16406 belongs to run-rb43028fb35dd4e8d85e086df1ca573df.service.
Thu 2022-03-10 16:51:55 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Main process exited, code=exited, status=0/SUCCESS (success)
Thu 2022-03-10 16:51:55 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Changed running -> stop-sigterm
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: State 'stop-sigterm' timed out. Killing.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Killing process 16407 (perl) with signal SIGKILL.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Changed stop-sigterm -> final-sigterm
Thu 2022-03-10 16:53:26 PST Received SIGCHLD from PID 16407 (perl).
Thu 2022-03-10 16:53:26 PST Child 16407 (perl) died (code=killed, status=9/KILL)
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Child 16407 belongs to run-rb43028fb35dd4e8d85e086df1ca573df.service.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Control group is empty.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Failed with result 'timeout'.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Service will not restart (restart setting)
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Changed final-sigterm -> failed
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Unit entered failed state.
Thu 2022-03-10 16:53:26 PST run-rb43028fb35dd4e8d85e086df1ca573df.service: Consumed 7ms CPU time.
```

You can see that after bash, the MAINPID, systemd tried to send sigterm to all processes, waiting a timeout, then sending sigkill.

Based on the documentation 

> Processes will first be terminated via SIGTERM (unless the signal to send is changed via KillSignal= or RestartKillSignal=). Optionally, this is immediately followed by a SIGHUP (if enabled with SendSIGHUP=). If processes still remain after the main process of a unit has exited or the delay configured via the TimeoutStopSec= has passed, the termination request is repeated with the SIGKILL signal or the signal specified via FinalKillSignal=

I expected the behavior of mixed and control-group to be the same. This is not the case (at least on cgroupv2)

My change here is an attempt to make the documentation say:

Processes will first be terminated via SIGTERM. If processes still remain after:

- the main process of a unit has exited (applies to Killmodes: mixed)
- the delay configured via the TimeoutStopSec= has passed (applies to KillModes: control-group, mixed, process)

 the termination request is repeated with the SIGKILL signal or the signal specified via FinalKillSignal=


Does that make sense? Should I change the wording?

**Comment by keszybz** [comment]
Yeah, this describes the current implementation correctly, AFAICT. I'll merge the patch.

FWIW, I think the whole description of of `KillMode=` should be rewritten. Right now it mentions `control-group`, but does not explain what it actually means, and then talks in detail about `mixed`. It's rally hard to follow.

**Comment by poettering** [comment]
Hmm, if ExecStop= is /bin/true, then the SIGKILL will also be triggered immediately. This is pretty relevant, hence I am not convinced the new text really is helpful
